### PR TITLE
[Intel MKL] Convolution op perchannel  support

### PR DESCRIPTION
**Pull request by nammbash**
## Convolution_Perchannel_op

1. Adding operators to support convolution to be performed perchannel and provided the mkl implementation for the same.

        QuantizedConv2DPerChannel

2. Provided Unit test to test the implementation logic of the above mentioned op.

### IMPORTANT NOTE: 
Depends on Requantization perchannel OP PR: https://github.com/tensorflow/tensorflow/pull/25203. The above mentioned PR Needs to be merged before this!

### CODE REVIEW

**Comment by nammbash** [comment]
@rthadur @penpornk this PR is a follow up to the requantization op per channel,  https://github.com/tensorflow/tensorflow/pull/25203 that needs review and approval.

**Comment by alextp** [comment]
I think between @penpornk and @tatianashp you have better reviewers than me for this code.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR!

File: tensorflow/core/api_def/base_api/api_def_QuantizedConv2DPerChannel.pbtxt
Comment (nammbash) [reply tensorflow/core/api_def/base_api/api_def_QuantizedConv2DPerChannel.pbtxt]: Done!

File: tensorflow/core/api_def/base_api/api_def_QuantizedConv2DPerChannel.pbtxt
Comment (nammbash) [reply tensorflow/core/api_def/base_api/api_def_QuantizedConv2DPerChannel.pbtxt]: I see! 
Looks like more developers are unaware of this as more files are found with
`description: <<END than"The original input tensor."`
than with
`description: "The original input tensor."`

Nevertheless changes done. Looks cleaner that way.

File: tensorflow/core/api_def/base_api/api_def_QuantizedConv2DPerChannel.pbtxt
Comment (nammbash) [reply tensorflow/core/api_def/base_api/api_def_QuantizedConv2DPerChannel.pbtxt]: done

File: tensorflow/core/graph/mkl_layout_pass.cc
Comment (nammbash) [reply tensorflow/core/graph/mkl_layout_pass.cc]: done

File: tensorflow/core/kernels/BUILD
Comment (nammbash) [reply tensorflow/core/kernels/BUILD]: done

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Hmm I agree cleaner c++11 way. Code is not entirely mine. Will check internally with developer on this one with his preference for a cleaner solution.

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done! Git merge blunder!

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done!

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: done.

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: done

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Agreed! I am proposing a work internally, on a a good design on handling INT8, vs UINT8 vs BFLOAT16 Mapping as this feels a little clumpsy when I am looking at the code. Would be best addressed then. Added a ToDO for now.

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done!

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: done

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: done

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: Created a separate internal ticket to address this concern and added a todo.

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: done

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: done

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (nammbash) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: Pleasure!

File: tensorflow/core/ops/mkl_nn_ops.cc
Comment (nammbash) [reply tensorflow/core/ops/mkl_nn_ops.cc]: done

**Comment by nammbash** [comment]
@penpornk Sorry for the delay! Was pulled into other project and blog work for google. Finally got some time to finish this PR. I think I have addressed all the review comments. let me know.

**Comment by nammbash** [comment]
@penpornk ( and @tatianashp )
Sorry for the delay! Was pulled into other project and blog work for Intel(google request). Finally got some time to finish this PR. I think I have addressed all the review comments. let me know.
@rthadur can you please trigger the tests again. Highly appreciate it.

**Comment by rthadur** [comment]
@nammbash  could you please sign CLA

**Comment by nammbash** [comment]
@rthadur that CLA is because of @Penpornk and a new feature of github. You can look at the comment below for further explanation. 
https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461987678

**Comment by nammbash** [comment]
@rthadur copied below for your quick reference.

From @penpornk here: #25203 (comment)
@hgadig He already signed the CLA. The bot detected his CLA and assigned CLA: yes since the beginning of the PR. Even the message right now says the bot found the CLA for @nammbash. What makes the bot assign CLA: no is because some of the commits are his accepting my online one-line suggestions (GitHub's new feature). For those commits, the author's emails that GitHub automatically used are 38869685+nammbash@users.noreply.github.com and 38085909+penpornk@users.noreply.github.com, which are not recognized by the googlebot.

Googlebot wants us to verify that all the coauthors in this PR have signed the CLA. There are only two people contributing to this PR: @nammbash and I. So I think we are fine. :)

**Comment by rthadur** [comment]
> #25203

sound good

File: tensorflow/core/ops/mkl_nn_ops.cc
Comment (nammbash) [reply tensorflow/core/ops/mkl_nn_ops.cc]: @penpornk I am not exactly clear on what you would like to be done here! Can you please be a little more detailed on this suggestion and I can try and address this review comment.

**Comment by penpornk** [comment]
Sorry for my delay. Will take a look at this soon. 
Regarding CLA, the situation is different from https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461987678

Looking at the commit info, there are two contributors, @nammbash and @ashahba, both have signed their CLAs individually. Quoting googlebot,

> 😕 The bad news is that it appears that one or more commits were authored or co-authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.

Each of you need to confirm here that you are ok with your commits being contributed to this project. I'm setting the CLA to no until then. Thank you!

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (penpornk) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Thank you!

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (penpornk) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Got it. Thank you!
Nits: `TODO (` -> `TODO(` and `regularly.A` -> `regularly. A`

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (penpornk) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: Thank you! Appreciate that. 

Nits: `TODO nammbash:` -> `TODO(nammbash):`, `all Mkl tests)` -> `all MKL tests)`.

File: tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc
Comment (penpornk) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_perchannel_test.cc]: Thank you for changing it!

File: tensorflow/core/ops/mkl_nn_ops.cc
Comment (penpornk) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Yes, I meant refactoring it to a common function and passing appropriate parameters. :)

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes, and no worries about the timing! I piled up a lot of PRs from your side in the past two weeks too. Sorry about that. Will gradually get back to them.

File: tensorflow/core/ops/mkl_nn_ops.cc
Comment (penpornk) [reply tensorflow/core/ops/mkl_nn_ops.cc]: We can save for in a separate PR too (just for this, so it's easier to debug when something is wrong). You can just add a TODO.

**Comment by ashahba** [comment]
@penpornk Looks like the one commit from me within the list of commits has caused the `cla` to be marked as no.
I tried to add label but I don't have access so if you need my consent then sure I am perfectly fine with the PR incorporating that commit.

Please let me know if you need anything else from me to unblock the PR.
Thanks.

**Comment by nammbash** [comment]
@penpornk Done! Thank you for the quick reviews on the week of Google Tensorflow Dev Summit.
I got the changes done.
CLA confusion should be resolved with this https://github.com/tensorflow/tensorflow/pull/25506#issuecomment-470231069

Are the tests passing? Seems like the api tests are failing even wihthout new api changes. Not from my PR?

**Comment by nammbash** [comment]
@penpornk https://github.com/tensorflow/tensorflow/pull/25506#issuecomment-470263913
Friendly reminder!

**Review by penpornk — APPROVED** [review]
Thank you for the changes and for your patience! I'm sorry again about the delay!
I'll change the CLA to yes since you both confirmed that you are okay with your commits being in the PR.

**Comment by rthadur** [comment]
@nammbash can you please check failed test

**Comment by penpornk** [comment]
@nammbash Please update the API golden files (both v1 and v2) and include them in the PR.
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
Full error log (Ubuntu Python 2) [here](https://source.cloud.google.com/results/invocations/c5a13c41-f6ae-4820-af6b-0ff95f2c4bde/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/log).

**Review by penpornk — CHANGES_REQUESTED** [review]
@nammbash Please update the API golden files (both v1 and v2) and include them in the PR.
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
Full error log (Ubuntu Python 2) [here](https://source.cloud.google.com/results/invocations/c5a13c41-f6ae-4820-af6b-0ff95f2c4bde/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/log).

**Comment by nammbash** [comment]
@penpornk @rthadur https://github.com/tensorflow/tensorflow/pull/25506#pullrequestreview-213796078
Done! please verify

**Review by penpornk — COMMENTED** [review]
Thank you! The changes to tensorflow.summary.pbtxt seem unrelated to this PR. Let's see what the tests say.

**Comment by nammbash** [comment]
> Thank you! The changes to tensorflow.summary.pbtxt seem unrelated to this PR. Let's see what the tests say.

I know! I mentioned it earlier here that the test failures are not related to my PR here: https://github.com/tensorflow/tensorflow/pull/25506#issuecomment-470263913

**Comment by penpornk** [comment]
@nammbash We still need the changes to `tensorflow.raw_ops.pbtxt` anyway (which are related to your PR because it adds `QuantizedConv2DPerChannel`). :) But I'm curious if we can remove the changes to `tensorflow.summary.pbtxt`.

**Comment by nammbash** [comment]
@penpornk I see! Anyways let me know! and thank you for your quick responses.

**Comment by nammbash** [comment]
@penpornk @rthadur Can you accept the cla so that the tests are triggered please!

**Comment by nammbash** [comment]
@rthadur @penpornk I suppose All tests have failed and the ones below are unrelated to my PR? Merge-able now?

Failures unrelated to my PR 
MacOS Contrib 
Windows Bazel GPU

Not sure about this
import/copybara

**Review by penpornk — CHANGES_REQUESTED** [review]
It seems `tensorflow.summary.pbtxt` really shouldn't be here. Could you please remove it from the PR? (But please keep `tensorflow.raw_ops.pbtxt` in the PR. We need them.) Thank you!

**Comment by nammbash** [comment]
@penpornk Done!. Accept the CLA and run the tests please!

**Review by penpornk — APPROVED** [review]
@nammbash Thank you for the quick fix! 

I have verified that the CLA is fine. I'm approving the PR for the non-API changes. We'll need the approval from API reviewers before we can go through. The API review (tagged) usually takes some time.

**Comment by nammbash** [comment]
@penpornk @rthadur Thank you for the quick responses. (I take it there is nothing from my end to do anymore, unless API team pings again)
https://github.com/tensorflow/tensorflow/pull/25506#pullrequestreview-214862831
We actually have 2 other internal PRs waiting on this to merge and hence pending.
Is there someone I can friendly request from the API team for a quick review? Much appreciated!

**Review by penpornk — DISMISSED** [review]
@nammbash I think they only hold a meeting to review all API changes once a week. @alextp can say more about this. 

In the meanwhile, please fix the API compatibility test failures. You can see the error in Ubuntu Python 2 [log](https://source.cloud.google.com/results/invocations/a9629395-eab8-4669-94bc-20ca42c61076/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/log).
```
   member_method {
     name: "QuantizedConv2DPerChannel"
-    argspec: "args=[\'input\', \'filter\', \'min_input\', \'max_input\', \'min_filter\', \'max_filter\', \'out_type\', \'strides\', \'padding\', \'dilations\'], varargs=None, keywords=None, defaults=None"
+    argspec: "args=[\'input\', \'filter\', \'min_input\', \'max_input\', \'min_filter\', \'max_filter\', \'strides\', \'padding\', \'out_type\', \'dilations\', \'name\'], varargs=None, keywords=None, defaults=[\"<dtype: \'qint32\'>\", \'[1, 1, 1, 1]\', \'None\'], "
   }
```

**Comment by nammbash** [comment]
@penpornk That is interesting. All these tests passed yesterday!? before in doing anything the summary.pbtxt

**Comment by penpornk** [comment]
@nammbash I left the tests running yesterday and hadn't checked the results until I last commented, so these errors are probably from yesterday. They are from `tensorflow.raw_ops.pbtxt` and are not from removing `tensorflow.summary.pbtxt`.

**Comment by nammbash** [comment]
@penpornk https://github.com/tensorflow/tensorflow/pull/25506#issuecomment-473475183
The failures are unrelated to my PR, the failures are seen in the master branch even without my PR changes.
(did a quick check and the failures seem to exist even n Feb 26 commit- But do verify this from your end.)
But as of this morning, the master branch in itself has the failures with respect to running the test and the failures seem to be unrelated to my PR:
```
bazel --output_base=../testmaster build --cache_test_results=no tensorflow/tools/api/tests:api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
```
@alextp to verify API friendly ping.

**Comment by nammbash** [comment]
@penpornk Yup! I tried to get the fresh public master and ran the command below:
```
bazel --output_base=../testmaster build --cache_test_results=no tensorflow/tools/api/tests:api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
```
it fails even without my PR Changes. Can you verify this and pull my PR in? or let me know otherwise?

**Comment by penpornk** [comment]
@nammbash I'm pretty sure the error is related to this PR, because it involves the API of `QuantizedConv2DPerChannel` which has just been introduced in this PR. Please see my comment https://github.com/tensorflow/tensorflow/pull/25506#pullrequestreview-215283956 for more details. 

We have pulled this inside and I have fixed the golden files. It's going through internal review process.

**Review by penpornk — COMMENTED** [review]
@nammbash I'm pretty sure the error is related to this PR, because it involves the API of `QuantizedConv2DPerChannel` which has just been introduced in this PR. Please see my comment https://github.com/tensorflow/tensorflow/pull/25506#pullrequestreview-215283956 for more details. 

We have pulled this inside and I have fixed the golden files. It's going through internal review process.

**Comment by nammbash** [comment]
@penpornk https://github.com/tensorflow/tensorflow/pull/25506#pullrequestreview-216077162
Aah! Got it! I did update the files earlier, but was confused as the api test was still failing even after the changes(as with the master).
Anyways, I think you hinted you changed it internally and is going through reviews. I have updated the files just in case.

Let me know!

**Comment by penpornk** [comment]
@nammbash Thanks for changing, but yes I've fixed it internally. The PR has been merged internally and the status here should change soon.

**Comment by penpornk** [comment]
I'm also setting the CLA to yes again (doesn't really matter now since we are not pulling more updates in, but just in case). Both contributors to this PR, @nammbash and @ashahba have said they are fine with their commits being used in the PR.

@rthadur The bottom of the page says this branch has a merging conflict (files out of date). I don't think this needs to be fixed since we are not pulling any new changes in. I had fixed the conflicts internally and the PR has been merged.

**Comment by penpornk** [comment]
The changes are already in https://github.com/tensorflow/tensorflow/commit/f949ead8b27a948b8319369875c7f048644dd194 so I'm closing this PR now.

**Comment by nammbash** [comment]
@penpornk Thank you for Merging and closing this PR. internally we have a question.
The Clang format does not seem to be pulled into the main branch and we happen to do the clang formatting again for unrelated PRs.
Specifically: (see clang format differences the PR had in Intel public branch)
see Intel-public branch here: https://github.com/Intel-tensorflow/tensorflow/blob/intel-mkl-483/tensorflow/core/kernels/mkl_conv_ops.cc#L464-L468
vs
Google public branch here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/mkl_conv_ops.cc#L464-L468

Any Idea what is happening. We are wondering internally!

**Pull request closed without merging**
